File: mobile_app/lib/utils/ref_utils.dart
/// Reference utilities to keep behavior consistent across the app.
/// - Preserve original (possibly zero-padded) reference strings for display
/// - Normalize (trim) for comparisons
/// - Provide numeric extraction for sorting only
class RefUtils {
  /// Normalize a reference by trimming whitespace. Keeps as string.
  static String normalizeRefString(String? ref) {
    return (ref ?? '').trim();
  }

  /// Extract numeric value from a reference string for sorting.
  /// Returns null if no digits are found.
  static int? numericRef(String? ref) {
    final s = normalizeRefString(ref);
    final match = RegExp(r'\d+').firstMatch(s);
    if (match != null) {
      return int.tryParse(match.group(0)!);
    }
    return int.tryParse(s);
  }

  /// Comparator for reference strings using numeric value when possible.
  static int compareRefStrings(String a, String b) {
    final na = numericRef(a);
    final nb = numericRef(b);
    if (na != null && nb != null) {
      return na.compareTo(nb);
    }
    return normalizeRefString(a).compareTo(normalizeRefString(b));
  }

  /// Format numeric references to a fixed width with leading zeros.
  /// If not purely numeric, returns the normalized input.
  static String formatRefWidth(String ref, {int width = 4}) {
    final s = normalizeRefString(ref);
    final numericOnly = RegExp(r'^\d+$');
    if (numericOnly.hasMatch(s)) {
      return s.padLeft(width, '0');
    }
    return s;
  }
}

File: desktop_matching_app/src/utils/refUtils.js
/**
 * Reference utilities shared across Electron apps.
 * Preserve original, zero-padded strings; use numeric extraction for sorting only.
 */

function normalizeRefString(ref) {
  if (ref == null) return '';
  return String(ref).trim();
}

function toNumericRef(ref) {
  const s = normalizeRefString(ref);
  const m = s.match(/\d+/);
  return m ? parseInt(m[0], 10) : Number.NaN;
}

function compareByNumericRef(a, b) {
  const na = toNumericRef(a?.Reference);
  const nb = toNumericRef(b?.Reference);
  if (Number.isNaN(na) && Number.isNaN(nb)) {
    return normalizeRefString(a?.Reference).localeCompare(normalizeRefString(b?.Reference));
  }
  if (Number.isNaN(na)) return 1;
  if (Number.isNaN(nb)) return -1;
  return na - nb;
}

function sortByNumericRef(items) {
  return items.slice().sort(compareByNumericRef);
}

function formatRefWidth(ref, width = 4) {
  const s = normalizeRefString(ref);
  if (/^\d+$/.test(s)) {
    return s.padStart(width, '0');
  }
  return s;
}

module.exports = {
  normalizeRefString,
  toNumericRef,
  compareByNumericRef,
  sortByNumericRef,
  formatRefWidth,
};
